### ISSUE TITLE
Quantizing GPT2-XL Model leads to "exceeds maximum protobuf"

### ISSUE THREAD

**Issue by collinarnett** [microsoft/onnxruntime#7017]
**Describe the bug**
GPT2-XL cannot be quantized when converting to int8

`ValueError: Message onnx.ModelProto exceeds maximum protobuf size of 2GB: 6552187990`

**Urgency**
I would like to get this done by the end of the week.

**System information**
- Ubuntu 18.04
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.7.0
- Python version: 3.8.8
- CUDA/cuDNN version: 11.1
- GPU model and memory: NA

**To Reproduce**
- Describe steps/code to reproduce the behavior.
Using a fine-tuned gpt2-xl model cannot be quantized with the example code on the transformers section of this repo:
```
python -m onnxruntime.transformers.convert_to_onnx -e -m pretrained --output quantized -p int8 --optimize_onnx
```
Where `pretrained` is the folder that contains my fine-tuned model that has been saved with `save_pretrained` from huggingface.  
- Attach the ONNX model to the issue (where applicable) to expedite investigation.
Cannot attach such a large mode

**Expected behavior**

**Screenshots**
```
Arguments:Namespace(cache_dir='./cache_models', input_test_file='', model_class='GPT2LMHeadModel', model_name_or_path='pretrained', optimize_onnx=True, output='quantized', precision=<Precision.INT8: 'int8'>, tolerance=3.0, use_external_data_format=True, use_gpu=False, verbose=False)
PyTorch Version:1.7.1+cu110
Transformers Version:4.3.2
Onnxruntime Version:1.7.0
Exporting ONNX model to quantized/pretrained_past/pretrained_past.onnx
Shapes: input_ids=torch.Size([1, 1]) past=torch.Size([2, 1, 25, 1, 64]) output=torch.Size([1, 1, 50257]) present=torch.Size([2, 1, 25, 2, 64])
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:654: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert batch_size > 0, "batch_size has to be defined and > 0"
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:169: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  w = w / (float(v.size(-1)) ** 0.5)
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:174: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  mask = self.bias[:, :, ns - nd : ns, :ns]
Optimizing model to quantized/pretrained_past_fp32/pretrained_past_fp32.onnx
Fused LayerNormalization count: 97
Fused FastGelu count: 48
Removed Reshape and Expand count: 0
Fused Attention(with past) count: 48
Graph pruned: 0 inputs, 0 outputs and 3129 nodes are removed
Position embedding path is not found. Embed layer cannot be fused.
Graph pruned: 0 inputs, 0 outputs and 1248 nodes are removed
postprocess: remove Reshape count:192
Fused FastGelu(add bias) count: 48
opset verion: 11
Output model to quantized/pretrained_past_fp32/pretrained_past_fp32.onnx
quantizing model...
Size of full precision ONNX model(MB):0.13769149780273438
onnxruntime.quantization.quantize is deprecated.
         Please use quantize_static for static quantization, quantize_dynamic for dynamic quantization.
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/convert_to_onnx.py", line 252, in <module>
    main()
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/convert_to_onnx.py", line 175, in main
    QuantizeHelper.quantize_onnx_model(output_path, onnx_model_paths['int8'], args.use_external_data_format)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/quantize_helper.py", line 75, in quantize_onnx_model
    quantized_onnx_model = quantize(
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/quantization/quantize.py", line 126, in quantize
    quantizer = ONNXQuantizer(copy_model, per_channel, nbits == 7, mode, static, symmetric_weight,
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/quantization/onnx_quantizer.py", line 32, in __init__
    model = onnx.shape_inference.infer_shapes(model)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnx/shape_inference.py", line 34, in infer_shapes
    model_str = model.SerializeToString()
ValueError: Message onnx.ModelProto exceeds maximum protobuf size of 2GB: 6552187990
```
I assume this is because https://github.com/microsoft/onnxruntime/blob/f4acdb2ecd254b17b3156baaba723873e2ef8c8a/onnxruntime/python/tools/quantization/quantize.py#L56 does not handle `use_external_data_format` however I changed 
https://github.com/microsoft/onnxruntime/blob/27ac88201aafc67bbd8ee439bb9bef745a4cc549/onnxruntime/python/tools/transformers/quantize_helper.py#L60
to use `quantize_dynamic` instead of the depreciated `quantize` but this did not resolve the error. There must be some missing handling of this large model.

**Additional context**
I just want to quantize my gpt2-xl model for faster inference.

**Comment by yufenglee** [comment]
@collinarnett, it is because onnx.shape_inference.infer_shapes doesn't support model > 2G. Will make a fix.

**Comment by collinarnett** [comment]
Hey I was just wondering if there was a status update on this. I am currently blocked on my project because of this issue.

**Comment by yufenglee** [comment]
It is in our backlog. Will try fixing the issue this week.

**Comment by yufenglee** [comment]
@collinarnett, you can work around this issue by comment out this line: https://github.com/microsoft/onnxruntime/blob/21c282ed542e7673d1eea0f54e113b84c40d5c50/onnxruntime/python/tools/quantization/onnx_quantizer.py#L32 , while I’m preparing a fix for the issue.

**Comment by collinarnett** [comment]
Hm, unfortunately when I run the same command as above with this line commented out I get the following error:
```
PyTorch Version:1.8.1+cu111
Transformers Version:4.5.0
Onnxruntime Version:1.7.0
Exporting ONNX model to quantized/pretrained_past/pretrained_past.onnx
Shapes: input_ids=torch.Size([1, 1]) past=torch.Size([2, 1, 25, 1, 64]) output=torch.Size([1, 1, 50257]) present=torch.Size([2, 1, 25, 2, 64])
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:659: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert batch_size > 0, "batch_size has to be defined and > 0"
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:170: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  w = w / (float(v.size(-1)) ** 0.5)
Optimizing model to quantized/pretrained_past_fp32/pretrained_past_fp32.onnx
Fused LayerNormalization count: 97
Fused FastGelu count: 48
Removed Reshape and Expand count: 0
Fused Attention(with past) count: 48
Graph pruned: 0 inputs, 0 outputs and 3129 nodes are removed
Position embedding path is not found. Embed layer cannot be fused.
Graph pruned: 0 inputs, 0 outputs and 1248 nodes are removed
postprocess: remove Reshape count:192
Fused FastGelu(add bias) count: 48
opset verion: 11
Output model to quantized/pretrained_past_fp32/pretrained_past_fp32.onnx
quantizing model...
Size of full precision ONNX model(MB):0.13768672943115234
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/convert_to_onnx.py", line 252, in <module>
    main()
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/convert_to_onnx.py", line 175, in main
    QuantizeHelper.quantize_onnx_model(output_path, onnx_model_paths['int8'], args.use_external_data_format)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/transformers/quantize_helper.py", line 63, in quantize_onnx_model
    quantized_onnx_model = quantize_dynamic(onnx_model_path,
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/quantization/quantize.py", line 283, in quantize_dynamic
    quantizer.model.save_model_to_file(model_output, use_external_data_format)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnxruntime/quantization/onnx_model.py", line 204, in save_model_to_file
    onnx.save_model(self.model, output_path)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnx/__init__.py", line 188, in save_model
    proto = write_external_data_tensors(proto, basepath)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnx/external_data_helper.py", line 253, in write_external_data_tensors
    save_external_data(tensor, filepath)
  File "/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/onnx/external_data_helper.py", line 161, in save_external_data
    open(external_data_file_path, 'ab').close()
FileNotFoundError: [Errno 2] No such file or directory: '/home/collin/litai/serving/quantized/pretrained_past_int8/pretrained_past_int8.onnx.data'
```

**Comment by collinarnett** [comment]
#6867 seem to fix that issue actually :+1:

**Comment by Narsil** [comment]
Wasn't this already an issue a few months ago ? https://github.com/microsoft/onnxruntime/pull/5210/files

**Comment by yaysummeriscoming** [comment]
> @collinarnett, you can work around this issue by comment out this line:
> 
> https://github.com/microsoft/onnxruntime/blob/21c282ed542e7673d1eea0f54e113b84c40d5c50/onnxruntime/python/tools/quantization/onnx_quantizer.py#L32
> 
> , while I’m preparing a fix for the issue.

This together with the nightly build that incorporates the fix @collinarnett suggested fixed things for me. Looking forward to a permanent fix!

**Comment by collinarnett** [comment]
> Wasn't this already an issue a few months ago ? https://github.com/microsoft/onnxruntime/pull/5210/files

Seems like it got added back in #6541 for some reason. 

---

On another note, quantizing gpt2-xl even from the base pretrained weights yields huge diffs between the pytorch model. Not sure if this is related to the shapes issue. @yufenglee 
```
python -m onnxruntime.transformers.convert_to_onnx -e -m "gpt2-xl" --output quantized -p int8 --optimize_onnx --tolerance 1e-6
Arguments:Namespace(cache_dir='./cache_models', input_test_file='', model_class='GPT2LMHeadModel', model_name_or_path='gpt2-xl', optimize_onnx=True, output='quantized', precision=<Precision.INT8: 'int8'>, tolerance=1e-06, use_external_data_format=True, use_gpu=False, verbose=False)
PyTorch Version:1.8.1+cu111
Transformers Version:4.5.0
Onnxruntime Version:1.7.0
Lock 140664125388400 acquired on ./cache_models/d2de8fec009fa9b9196047559bcac6c1f02a9c500718b4346bc516354965b1ca.81d9c13b9ee3f2b22faaba04ca49e09b13f9fea3a7910768ed6664ec141e3c8b.lock
Downloading: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 787/787 [00:00<00:00, 584kB/s]
Lock 140664125388400 released on ./cache_models/d2de8fec009fa9b9196047559bcac6c1f02a9c500718b4346bc516354965b1ca.81d9c13b9ee3f2b22faaba04ca49e09b13f9fea3a7910768ed6664ec141e3c8b.lock
Lock 140663701315936 acquired on ./cache_models/96569b907e56747ce3e593c6a13d8475b8c733a64aab8af8f602b90d94c4af71.8fbbcdf404c82c5967934d411f1462fa0574d639f2aa398aa3754fced1bb26c0.lock
Downloading: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6.43G/6.43G [02:42<00:00, 39.5MB/s]
Lock 140663701315936 released on ./cache_models/96569b907e56747ce3e593c6a13d8475b8c733a64aab8af8f602b90d94c4af71.8fbbcdf404c82c5967934d411f1462fa0574d639f2aa398aa3754fced1bb26c0.lock
Exporting ONNX model to quantized/gpt2-xl_past/gpt2-xl_past.onnx
Shapes: input_ids=torch.Size([1, 1]) past=torch.Size([2, 1, 25, 1, 64]) output=torch.Size([1, 1, 50257]) present=torch.Size([2, 1, 25, 2, 64])
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:659: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert batch_size > 0, "batch_size has to be defined and > 0"
/home/collin/.virtualenvs/litai/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py:170: TracerWarning: Converting a tensor to a Python float might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  w = w / (float(v.size(-1)) ** 0.5)
Optimizing model to quantized/gpt2-xl_past_fp32/gpt2-xl_past_fp32.onnx
Fused LayerNormalization count: 97
Fused FastGelu count: 48
Removed Reshape and Expand count: 0
Fused Attention(with past) count: 48
Graph pruned: 0 inputs, 0 outputs and 3129 nodes are removed
Position embedding path is not found. Embed layer cannot be fused.
Graph pruned: 0 inputs, 0 outputs and 1248 nodes are removed
postprocess: remove Reshape count:192
Fused FastGelu(add bias) count: 48
opset verion: 11
Output model to quantized/gpt2-xl_past_fp32/gpt2-xl_past_fp32.onnx
quantizing model...
Size of full precision ONNX model(MB):0.1353445053100586
onnxruntime.quantization.quantize is deprecated.
         Please use quantize_static for static quantization, quantize_dynamic for dynamic quantization.
quantized model saved to:quantized/gpt2-xl_past_int8/gpt2-xl_past_int8.onnx
Size of quantized ONNX model(MB):0.28689098358154297
Size of full precision Torch model(MB):5990.049831390381
Size of quantized Torch model(MB):1848.1334533691406
finished quantizing model
Output path: quantized/gpt2-xl_past_int8/gpt2-xl_past_int8.onnx
Running parity test (rtol=1e-06, atol=1e-06, test_cases=100, use_io_binding=True model_class=GPT2LMHeadModel is_float16=False) ...
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.90612
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.27802
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.48389
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.32870
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.95715
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.46044
PyTorch and OnnxRuntime results are not all close: max_abs_diff=9.24362
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.39315
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.52452
PyTorch and OnnxRuntime results are not all close: max_abs_diff=8.76055
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.75484
PyTorch and OnnxRuntime results are not all close: max_abs_diff=2.66290
PyTorch and OnnxRuntime results are not all close: max_abs_diff=2.85430
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.41245
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.02360
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.67525
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.63748
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.55965
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.60409
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.89149
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.29298
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.44962
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.17730
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.19035
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.17519
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.58201
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.09830
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.57839
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.70320
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.85671
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.65241
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.41488
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.96500
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.25346
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.45235
PyTorch and OnnxRuntime results are not all close: max_abs_diff=2.46278
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.00452
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.50152
PyTorch and OnnxRuntime results are not all close: max_abs_diff=8.33231
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.80440
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.40989
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.40426
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.03675
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.22421
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.82677
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.52984
PyTorch and OnnxRuntime results are not all close: max_abs_diff=2.22594
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.16457
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.52313
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.49823
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.87267
PyTorch and OnnxRuntime results are not all close: max_abs_diff=10.36788
PyTorch and OnnxRuntime results are not all close: max_abs_diff=10.87117
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.00919
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.01933
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.89842
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.83284
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.51561
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.46636
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.75223
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.35438
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.25203
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.52229
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.50336
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.86641
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.45779
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.89096
PyTorch and OnnxRuntime results are not all close: max_abs_diff=4.50498
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.04193
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.13582
PyTorch and OnnxRuntime results are not all close: max_abs_diff=8.16340
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.81378
PyTorch and OnnxRuntime results are not all close: max_abs_diff=2.67974
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.23957
PyTorch and OnnxRuntime results are not all close: max_abs_diff=9.00699
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.85490
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.51197
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.15913
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.21516
PyTorch and OnnxRuntime results are not all close: max_abs_diff=8.69954
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.99599
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.22929
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.35550
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.36479
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.17758
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.28299
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.73311
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.20211
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.15749
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.71275
PyTorch and OnnxRuntime results are not all close: max_abs_diff=3.70586
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.50874
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.46755
PyTorch and OnnxRuntime results are not all close: max_abs_diff=7.94581
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.88780
PyTorch and OnnxRuntime results are not all close: max_abs_diff=10.62353
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.98177
PyTorch and OnnxRuntime results are not all close: max_abs_diff=6.26682
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.91176
PyTorch and OnnxRuntime results are not all close: max_abs_diff=5.68204
Parity Test Cases=100; Passed=0
Done. Output model: quantized/gpt2-xl_past_int8/gpt2-xl_past_int8.onnx
```

Not sure if this makes sense since the example in the gpt2 notebook shows a very small difference in the logits.

**Comment by whiteRa2bit** [comment]
Are there any updates regarding the issue?
I am also trying to convert DialoGPT-large model to onnx,

```
python -m onnxruntime.transformers.convert_to_onnx \
        -m model_path \
        --model_class GPT2LMHeadModel \
        --output /root/models/gpt_onnx \
        --use_external_data_format \
        -p fp16 \
        --optimize_onnx \
        --use_gpu
```

The output of the script is the following:
```
PyTorch Version:1.9.0a0+2ecb2c7
Transformers Version:4.6.1
Onnxruntime Version:1.8.0
....
  File "/opt/conda/lib/python3.8/site-packages/onnxconverter_common/float16.py", line 130, in convert_float_to_float16
    model = func_infer_shape(model)
  File "/opt/conda/lib/python3.8/site-packages/onnx/shape_inference.py", line 34, in infer_shapes
    model_str = model.SerializeToString()
ValueError: Message onnx.ModelProto exceeds maximum protobuf size of 2GB: 3391372453
```

The conversion works when I comment the line 130 in `onnxconverter_common/float16.py` script, but the quality of the generation of the converted model is noticeably lower.

**Comment by leoozy** [comment]
I met the same problems while quantizating the mBart model (about 3GB)

**Comment by leoozy** [comment]
@snnn

**Comment by yufenglee** [comment]
We just added a option in quantization to disable shapeinference in quantization. You can enable this option to solve this issue: https://github.com/microsoft/onnxruntime/blob/60089f7093e4e26f837be4bbf74d38cb97b43e4b/onnxruntime/python/tools/quantization/quantize.py#L189

**Comment by collinarnett** [comment]
How come there is such a drastic difference in logits when shape inference is turned off as I referenced in #6549? @yufenglee

**Comment by yufenglee** [comment]
I don't think it is related to shape inference. To confirm this, could you try calling this API to do shape inference before calling quantizatize_dynamic: https://github.com/onnx/onnx/blob/5515769e1950ce0bcf5f75cfc336a5c96a56b798/docs/PythonAPIOverview.md#shape-inference-a-large-onnx-model-2gb? 
Quantization can introduce accuracy loss. For very large model, the accuracy loss can be aggregated. 

`
onnx.shape_inference.infer_shapes_path('path/to/the/model.onnx', 'output/inferred/model.onnx')

`

**Comment by yufenglee** [comment]
Fixed in PR #11106
